### ISSUE TITLE
Error with running neural_network.py

### ISSUE THREAD

**Issue by pje1217** [ArielLevineLabNINDS/SeqSeek_Classify_Full_Pipeline#1]
[2021-11-08_SeqSeek.32824690.log](https://github.com/ArielLevineLabNINDS/SeqSeek_Classify_Full_Pipeline/files/7515738/2021-11-08_SeqSeek.32824690.log)

Hello, Thank you for developing a great resource and tool for the spinal cord scRNAseq community. 
I was able to run the first part of SeqSeek and got coarse cell types. 
However, I am having issues running neural_network.py
I am not sure what the issue is. I believe I installed everything I need by 
$ pip install -r requirements.txt

Also, although I got 4 outputs after "label_transfer.R", my output file recorded saving coarse results as NULL. 

Predicting cell labels
	Saving coarse results...
NULL

I was wondering if I am getting an error with "neural_network.py" because there is an error in the output file from "label_transfer.R"?

I would appreciate your help. Thank you!

**Comment by danielruss** [comment]
Hi,
  Looking at your log it appears that the code requires a nvidia GPU which you don't appear to have.  

```
2021-11-09 13:43:05.497475: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
```
@rbpatt2019  Do we have a non cuda version of the code?

**Comment by pje1217** [comment]
Thank you so much for your prompt reply. 
I am new to python and all the environment required for python. Could you tell me how to install nvidia GPU?
I am processing the data via the university server. If it's complicated to install nvidia GPU and if you have a non cuda version of the code, that would be awesome. Thank you!

**Comment by danielruss** [comment]
A GPU is a piece of hardware in your computer.  Your video card is a GPU.  However, it can perform matrix math super fast, so it is great for neural networks.  The neural network software we use (tensorflow) only makes use of NVIDIA gpu's.   I assumed you were running on a laptop/desktop. However, if you are running on a university server, you should ask your IT department if you access to a server with a GPU. 

Otherwise, hopefully  Ryan will get back to us.  He's in England, so we may not hear from him until tomorrow.  Sadly, tomorrow is a US Federal Holiday (Veteran's Day), so I wont be here tomorrow to see if he responds.

**Comment by danielruss** [comment]
Did you see https://github.com/ArielLevineLabNINDS/SeqSeek_Classify_NeuralNet_Only

**Comment by pje1217** [comment]
I see. Thank you for your help Daniel. I will check out the link you shared. I haven't looked into it yet.

**Comment by jieunesther** [comment]
Hi Daniel, 

I am following up on my previous question. I apologize for the change in my account. 
I was able to have access to a server with a GPU and it seems to load libcuda.so.1 now, but I still get an error with loading other libraries. 
Does it mean I will have to download these libraries first in a server with a GPU before running the code?
Below is the output file I got. 

Thank you for your help!

Best,
Esther






 
The following have been reloaded with a version change:
  1) r/4.1.0 => r/3.6.0

Running stage 'NN' with command:
	python src/py/neural_network.py
2021-11-11 16:50:03.402320: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2021-11-11 16:50:03.465562: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:1a:00.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-11-11 16:50:03.465829: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2021-11-11 16:50:03.465977: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory
2021-11-11 16:50:03.558527: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2021-11-11 16:50:03.670000: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2021-11-11 16:50:03.670250: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory
2021-11-11 16:50:03.670376: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory
2021-11-11 16:50:03.876749: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2021-11-11 16:50:03.876785: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2021-11-11 16:50:03.878713: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2021-11-11 16:50:03.906926: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 2400000000 Hz
2021-11-11 16:50:03.907795: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x4aa9240 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2021-11-11 16:50:03.907827: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2021-11-11 16:50:03.910518: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-11-11 16:50:03.910539: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      
Reading count matrix
Loading model
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense (Dense)                (None, 256)               6507520   
_________________________________________________________________
cell_type (Dense)            (None, 70)                17990     
=================================================================
Total params: 6,525,510
Trainable params: 6,525,510
Non-trainable params: 0
_________________________________________________________________
Making predictions
Traceback (most recent call last):
  File "src/py/neural_network.py", line 45, in <module>
    preds = np.argmax(model.predict(counts), axis=1)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 88, in _method_wrapper
    return method(self, *args, **kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1268, in predict
    tmp_batch_outputs = predict_function(iterator)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 580, in __call__
    result = self._call(*args, **kwds)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 627, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 506, in _initialize
    *args, **kwds))
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2446, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2777, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2667, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/framework/func_graph.py", line 981, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 441, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in user code:

    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py:1147 predict_function  *
        outputs = self.distribute_strategy.run(
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:951 run  **
        return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:2290 call_for_each_replica
        return self._call_for_each_replica(fn, args, kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:2649 _call_for_each_replica
        return fn(*args, **kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py:1122 predict_step  **
        return self(x, training=False)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py:886 __call__
        self.name)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/input_spec.py:216 assert_input_compatibility
        ' but received input with shape ' + str(shape))

    ValueError: Input 0 of layer sequential_6 is incompatible with the layer: expected axis -1 of input shape to have value 25419 but received input with shape [None, 22798]

ERROR: failed to reproduce 'dvc.yaml': failed to run: python src/py/neural_network.py, exited with 1

**Comment by jieunesther** [comment]
Hi Daniel, following up on my previous issue reported- I am now able to load the libraries using a server with the access to NVIDIA gpu. Thank you!

But I still get an error with running the code, and I am not sure what the issue is. 
My inputs for the neural_network.py are the files outputted after label_transfer.R 
Below is the output log file. 





Reading count matrix
Loading model
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense (Dense)                (None, 256)               6507520   
_________________________________________________________________
cell_type (Dense)            (None, 70)                17990     
=================================================================
Total params: 6,525,510
Trainable params: 6,525,510
Non-trainable params: 0
_________________________________________________________________
Making predictions
Traceback (most recent call last):
  File "src/py/neural_network.py", line 45, in <module>
    preds = np.argmax(model.predict(counts), axis=1)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 88, in _method_wrapper
    return method(self, *args, **kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1268, in predict
    tmp_batch_outputs = predict_function(iterator)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 580, in __call__
    result = self._call(*args, **kwds)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 627, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 506, in _initialize
    *args, **kwds))
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2446, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2777, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/function.py", line 2667, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/framework/func_graph.py", line 981, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/eager/def_function.py", line 441, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/framework/func_graph.py", line 968, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in user code:

    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py:1147 predict_function  *
        outputs = self.distribute_strategy.run(
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:951 run  **
        return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:2290 call_for_each_replica
        return self._call_for_each_replica(fn, args, kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/distribute/distribute_lib.py:2649 _call_for_each_replica
        return fn(*args, **kwargs)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/training.py:1122 predict_step  **
        return self(x, training=False)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py:886 __call__
        self.name)
    /pine/scr/j/i/jieunp/SeqSeek_Pipeline/SeqSeek/lib64/python3.6/site-packages/tensorflow/python/keras/engine/input_spec.py:216 assert_input_compatibility
        ' but received input with shape ' + str(shape))

    ValueError: Input 0 of layer sequential_6 is incompatible with the layer: expected axis -1 of input shape to have value 25419 but received input with shape [None, 22798]

ERROR: failed to reproduce 'dvc.yaml': failed to run: python src/py/neural_network.py, exited with 1

**Comment by danielruss** [comment]
Hello,
   First, great job getting the GPUs.  The next problem is that the neural net expect a particular set of 25419 genes.  You only have 22798 genes in your data.  You will need to insert zeros for all the genes you don't have data.
  This is similar to the case where you have a different genome.  https://colab.research.google.com/github/ArielLevineLabNINDS/SeqSeek-Classify-NN/blob/master/example_use_NN.ipynb.   Then goto "So you want to run your data..."

**Comment by jieunesther** [comment]
Thank you Daniel for directing me to the next step!
I was able to generate a matrix with a set of 25419 genes following the instructions and now I am trying again. I hope it works this time!
Thank you for developing a very useful tool for the spinal cord community. I truly appreciate your support!
I want to point out that there was a type error in line 61 of "https://raw.githubusercontent.com/ArielLevineLabNINDS/SeqSeek-Classify-NN/master/util/RemapGenes.R" 

there are two places where it's written "countMatrix" when it has to be "count_matrix"

I want to point out one more thing for people who will utilize this tool for future:
my data is integrated and I had to change "label_transfer.R" script in SeqSeek_Pipeline/src/R/ folder. 
I had to revise FindTransferAnchors query.assay from "RNA" to "integrated" (like below) to make it work for my dataset. 
I thought it would be helpful if this is noted in the instruction page for future!

Thank you very much!!


anchors <- FindTransferAnchors(
  reference = reference,
  query = query,
  normalization.method = "LogNormalize",
  reference.assay = "integrated",
  query.assay = "integrated",
  reduction = "pcaproject",
  features = VariableFeatures(object = reference),
  npcs = NULL,
  dims = 1:28
)
